lib/Dialect/Arc/Transforms/StripSV.cpp

```diff
@@ -0,0 +1,178 @@
+#include "mlir/IR/ImplicitLocOpBuilder.h"
+#include "llvm/Support/Debug.h"
+#include <variant>
+
+#define DEBUG_TYPE "arc-strip-sv"
+
+using namespace circt;
+using namespace arc;
+
+namespace {
+struct StripSVPass : public StripSVBase<StripSVPass> {
+  void runOnOperation() override;
+  SmallVector<Operation *> opsToDelete;
+  SmallPtrSet<StringAttr, 4> clockGateModuleNames;
+  SmallPtrSet<StringAttr, 4> externModuleNames;
+};
+} // namespace
+
+void StripSVPass::runOnOperation() {
+  auto mlirModule = getOperation();
+  opsToDelete.clear();
+  clockGateModuleNames.clear();
+  externModuleNames.clear();
+
+  auto expectedClockGateInputs =
+      ArrayAttr::get(&getContext(), {StringAttr::get(&getContext(), "in"),
+                                     StringAttr::get(&getContext(), "test_en"),
+                                     StringAttr::get(&getContext(), "en")});
+  auto expectedClockGateOutputs =
+      ArrayAttr::get(&getContext(), {StringAttr::get(&getContext(), "out")});
+  auto i1Type = IntegerType::get(&getContext(), 1);
+
+  for (auto extModOp : mlirModule.getOps<hw::HWModuleExternOp>()) {
+    if (extModOp.getVerilogModuleName() == "EICG_wrapper") {
+      if (extModOp.getArgNames() != expectedClockGateInputs ||
+          extModOp.getResultNames() != expectedClockGateOutputs) {
+        extModOp.emitError("clock gate module `")
+            << extModOp.moduleName() << "` has incompatible port names "
+            << extModOp.getArgNamesAttr() << " -> "
+            << extModOp.getResultNamesAttr();
+        return signalPassFailure();
+      }
+      if (extModOp.getArgumentTypes() !=
+              ArrayRef<Type>{i1Type, i1Type, i1Type} ||
+          extModOp.getResultTypes() != ArrayRef<Type>{i1Type}) {
+        extModOp.emitError("clock gate module `")
+            << extModOp.moduleName() << "` has incompatible port types "
+            << extModOp.getArgumentTypes() << " -> "
+            << extModOp.getResultTypes();
+        return signalPassFailure();
+      }
+      clockGateModuleNames.insert(extModOp.moduleNameAttr());
+    } else {
+      externModuleNames.insert(extModOp.moduleNameAttr());
+    }
+    opsToDelete.push_back(extModOp);
+  }
+  LLVM_DEBUG(llvm::dbgs() << "Found " << clockGateModuleNames.size()
+                          << " clock gates, " << externModuleNames.size()
+                          << " other extern modules\n");
+
+  // Remove `sv.*` operation attributes.
+  mlirModule.walk([](Operation *op) {
+    auto isSVAttr = [](NamedAttribute attr) {
+      return attr.getName().getValue().startswith("sv.");
+    };
+    if (llvm::any_of(op->getAttrs(), isSVAttr)) {
+      SmallVector<NamedAttribute> newAttrs;
+      newAttrs.reserve(op->getAttrs().size());
+      for (auto attr : op->getAttrs())
+        if (!isSVAttr(attr))
+          newAttrs.push_back(attr);
+      op->setAttrs(newAttrs);
+    }
+  });
+
+  // Remove ifdefs and verbatim.
+  for (auto verb : mlirModule.getOps<sv::VerbatimOp>())
```

Comment:
That's bold.  I guess it might work at the top-level sometimes.  Seems very coupled to firrtl's lowering.